aten/src/ATen/core/function_schema.h

```diff
@@ -78,6 +95,38 @@ struct Argument {
     return Argument(name_, new_type, N_, default_value_, kwarg_only_, alias_info_);
   }
 
+  // this function check whether this Argument is backward compatible with
+  // the old_argument. we intent to be conservative, if necessary, we may
+  // relax the checks in future. we consider the following the situations are
+  // backward compatible:
+  //   1) two arguments are equal
+  //   2) old_argument's type is T, and this Argument's type is Optional[T]
+  //   3) old_argument has no default value, and this Argument provides default
+  //      value
+  bool isBackwardCompatibleWith(const Argument& old_argument) const {
```

Comment:
nit: there's function_schema_inl.h - maybe move implementation there (as it's lengthy)